crates/ruff_python_semantic/src/statements.rs

```diff
@@ -0,0 +1,100 @@
+/// `NonZeroU32` to take advantage of memory layout optimizations.
+#[newtype_index]
+#[derive(Ord, PartialOrd)]
+pub struct StatementId;
+
+/// A [`Stmt`] AST node, along with a pointer to its parent statement (if any).
+#[derive(Debug)]
+struct Statement<'a> {
+    /// A pointer to the AST node.
+    node: &'a Stmt,
+    /// The ID of the parent of this node, if any.
+    parent: Option<StatementId>,
+    /// The depth of this node in the tree.
+    depth: u32,
+}
+
+/// The statements of a program indexed by [`StatementId`]
+#[derive(Debug, Default)]
+pub struct Statements<'a> {
+    nodes: IndexVec<StatementId, Statement<'a>>,
+    node_to_id: FxHashMap<RefEquality<'a, Stmt>, StatementId>,
+}
+
+impl<'a> Statements<'a> {
+    /// Inserts a new statement into the node tree and returns its unique ID.
+    ///
+    /// Panics if a statement with the same pointer already exists.
+    pub(crate) fn insert(&mut self, node: &'a Stmt, parent: Option<StatementId>) -> StatementId {
+        let next_id = self.nodes.next_index();
+        if let Some(existing_id) = self.node_to_id.insert(RefEquality(node), next_id) {
+            panic!("Node already exists with id {existing_id:?}");
+        }
+        self.nodes.push(Statement {
+            node,
+            parent,
+            depth: parent.map_or(0, |parent| self.nodes[parent].depth + 1),
+        })
+    }
+
+    /// Returns the [`StatementId`] of the given statement.
+    #[inline]
+    pub fn node_id(&self, statement: &'a Stmt) -> Option<StatementId> {
+        self.node_to_id.get(&RefEquality(statement)).copied()
+    }
+
+    /// Return the [`StatementId`] of the parent statement.
+    #[inline]
+    pub fn parent_id(&self, statement_id: StatementId) -> Option<StatementId> {
+        self.nodes[statement_id].parent
+    }
+
+    /// Return the parent of the given statement.
+    pub fn parent(&self, statement: &'a Stmt) -> Option<&'a Stmt> {
+        let statement_id = self.node_to_id.get(&RefEquality(statement))?;
+        let parent_id = self.nodes[*statement_id].parent?;
+        Some(self[parent_id])
+    }
+
+    /// Return the depth of the statement.
+    #[inline]
+    pub(crate) fn depth(&self, node_id: StatementId) -> u32 {
+        self.nodes[node_id].depth
+    }
+
+    /// Returns an iterator over all [`StatementId`] ancestors, starting from the given [`StatementId`].
+    pub(crate) fn ancestor_ids(
+        &self,
+        node_id: StatementId,
+    ) -> impl Iterator<Item = StatementId> + '_ {
+        std::iter::successors(Some(node_id), |&node_id| self.nodes[node_id].parent)
+    }
+}
+
+impl<'a> Index<StatementId> for Statements<'a> {
+    type Output = &'a Stmt;
+
+    #[inline]
+    fn index(&self, index: StatementId) -> &Self::Output {
+        &self.nodes[index].node
+    }
+}
+
+impl<'a> IndexMut<StatementId> for Statements<'a> {
+    #[inline]
+    fn index_mut(&mut self, index: StatementId) -> &mut Self::Output {
+        &mut self.nodes[index].node
+    }
+}
```

Comment:
Do we need the mutability (and indexing)? Same for `Expr`.

Reply:
No! Good call.